tests/test_frontend_testing_api/test_reservation.py

```diff
@@ -0,0 +1,151 @@
+
+pytestmark = [
+    pytest.mark.django_db,
+]
+
+
+_ENDPOINT = "frontend_testing_api:reservation-list"
+
+
+def test_frontend_api__reservation__user__general_role(api_client):
+    data = {
+        "user": {"role": "HANDLER"},
+        "params": {},
+    }
+    response = api_client.post(reverse(_ENDPOINT), data=data, format="json")
+    assert response.status_code == 201, response.json()
+
+    user = User.objects.first()
+    assert user.permissions.can_manage_reservation(Reservation.objects.first(), reserver_needs_role=True)
+    assert user.permissions.has_general_role(role_choices=["HANDLER"])
+    assert not user.permissions.has_role_for_units_or_their_unit_groups(units=Unit.objects.all())
+
+
+def test_frontend_api__reservation__user__unit_role(api_client):
+    data = {
+        "user": {"role": "ADMIN", "is_unit_role": True},
+        "params": {},
+    }
+    response = api_client.post(reverse(_ENDPOINT), data=data, format="json")
+    assert response.status_code == 201, response.json()
+
+    user = User.objects.first()
+    assert user.permissions.can_manage_reservation(Reservation.objects.first(), reserver_needs_role=True)
+    assert not user.permissions.has_general_role()
+    assert user.permissions.has_role_for_units_or_their_unit_groups(units=Unit.objects.all())
+
+
+def test_frontend_api__reservation__user__not_provided_no_auth(api_client):
+    data = {
+        "params": {},
+    }
+    response = api_client.post(reverse(_ENDPOINT), data=data, format="json")
+    assert response.status_code == 201, response.json()
+
+    user = User.objects.first()
+    reservation = Reservation.objects.first()
+    assert user == reservation.user
+    assert not user.permissions.can_manage_reservation(reservation, reserver_needs_role=True)
+    assert not user.permissions.has_general_role()
+    assert not user.permissions.has_role_for_units_or_their_unit_groups(units=Unit.objects.all())
+
+
+def test_frontend_api__reservation__set_datetime__default_options(api_client):
+    data = {
+        "current_datetime": "2022-01-01T12:00:00",
+        "user": {},
+        "params": {},
+    }
+    response = api_client.post(reverse(_ENDPOINT), data=data, format="json")
+    assert response.status_code == 201, response.json()
+
+    reservation_unit = ReservationUnit.objects.first()
+    assert reservation_unit.is_archived is False
+    assert reservation_unit.is_draft is False
+    assert reservation_unit.require_reservation_handling is False
+    assert reservation_unit.cancellation_rule.can_be_cancelled_time_before is None
+
+    reservation = Reservation.objects.first()
+    assert reservation.begin == local_datetime(2022, 1, 2, 12)
+    assert reservation.state == ReservationStateChoice.CONFIRMED
+    assert reservation.recurring_reservation is None
+
+    pricing = ReservationUnitPricing.objects.first()
+    assert pricing.highest_price == 0
+
+    assert PaymentOrder.objects.count() == 0
+
+    stop_time_freeze()
```

Comment:
This will not get run if the test runs into an error (like an assert failing), which would break the rest of the tests due to wrong local time begin set. Instead, use a decorator with a finally block:

```python
@contextmanager
def unfreeze_at_end():
    try:
        yield
    finally:
        stop_time_freeze()

@unfreeze_at_end
def test_frontend_api__reservation__set_datetime__default_options(api_client): ...
```